internal/config/loader.go

```diff
@@ -10,38 +11,107 @@ package config
+// should be used to read config settings. This also controls
 // whether the service registers itself the registry. The profile and confDir
 // are used to locate the local TOML config file.
-func LoadConfig(useRegistry bool, profile string, confDir string) (config *common.Config, err error) {
+func LoadConfig(useRegistry bool, profile string, confDir string) (*common.Config, error) {
 	fmt.Fprintf(os.Stdout, "Init: useRegistry: %v profile: %s confDir: %s\n",
 		useRegistry, profile, confDir)
-	confName := "configuration.toml"
 
+	configuration, err := loadConfigFromFile(profile, confDir)
+
+	// TODO: Verify this is correct.
+	stem := common.ConfigRegistryStem + common.ServiceName + "/"
+
+	var registryMsg string
+	if useRegistry {
+		registryMsg = "Register in registry..."
+		registryConfig := registry.Config{
+			Host:          configuration.Registry.Host,
+			Port:          configuration.Registry.Port,
+			Type:          configuration.Registry.Type,
+			Stem:          stem,
+			CheckInterval: configuration.Registry.CheckInterval,
+			CheckRoute:    common.APIPingRoute,
+			ServiceKey:    common.ServiceName,
+			ServiceHost:   configuration.Service.Host,
+			ServicePort:   configuration.Service.Port,
+		}
+
+		RegistryClient, err =  factory.NewRegistryClient(registryConfig)
+		if err != nil {
+			return nil, fmt.Errorf("connection to Registry could not be made: %v", err.Error())
+		}
+
+		// Check if registry service is running
+		if err := checkRegistryUp(configuration); err != nil {
+			return nil, err
+		}
+
+		// Register the service with Registry for discovery and health checks
+		err = RegistryClient.Register()
+		if err != nil {
+			return nil, fmt.Errorf("could not register service with Registry: %v", err.Error())
+		}
+
+		hasConfiguration, err := RegistryClient.HasConfiguration()
+		err = RegistryClient.Register()
+		if err != nil {
+			return nil, fmt.Errorf("could not verify that Registry already has configuration: %v", err.Error())
+		}
+
+		if hasConfiguration{
+			// Get the configuration values from the Registry
+			rawConfig, err := RegistryClient.GetConfiguration(configuration)
+			if err != nil {
+				return nil, fmt.Errorf("could not get configuration from Registry: %v", err.Error())
+			}
+
+			actual, ok := rawConfig.(*common.Config)
+			if !ok {
+				return nil, fmt.Errorf("configuration from Registry failed type check")
+			}
+
+			configuration = actual
+		} else {
+			// Self bootstrap the Registry with the device service's configuration
+			err := RegistryClient.PutConfiguration(*configuration, true)
+			if err != nil {
+				return nil, fmt.Errorf("could not push configuration to Registry: %v", err.Error())
+			}
+		}
+	} else {
+		registryMsg = "Bypassing registration in registry..."
+	}
+
+	fmt.Println(registryMsg)
+
+	return configuration, nil
+}
+
+func loadConfigFromFile(profile string, confDir string) (config *common.Config, err error) {
 	if len(confDir) == 0 {
 		confDir = "./res"
```

Comment:
it might be a good time to move this default confDir to common/consts.go along with the ConfigFileName

Reply:
Ok